lib/stats/stats-aggregator-register.c

```diff
@@ -0,0 +1,260 @@
+
+static GStaticMutex stats_aggregator_mutex = G_STATIC_MUTEX_INIT;
+static gboolean stats_aggregator_locked;
+static struct iv_timer update_timer;
+
+static void
+_update_func (gpointer _key, gpointer _value, gpointer _user_data)
+{
+  StatsAggregator *self = (StatsAggregator *) _value;
+  stats_aggregator_aggregate(self);
+}
+
+static void
+_start_timer(void)
+{
+  main_loop_assert_main_thread();
+  iv_validate_now();
+  update_timer.expires = iv_now;
+  update_timer.expires.tv_sec += FREQUENCY_OF_UPDATE;
+
+  iv_timer_register(&update_timer);
+}
+
+static void
+_update(void *cookie)
+{
+  g_hash_table_foreach(stats_cluster_container.clusters, _update_func, NULL);
+
+  if(g_hash_table_size(stats_cluster_container.clusters) > 0 && !iv_timer_registered(&update_timer))
+    _start_timer();
+}
+
+static void
+_init_timer(void)
+{
+  IV_TIMER_INIT(&update_timer);
+  update_timer.cookie = NULL;
+  update_timer.handler = _update;
+}
+
+static void
+_stop_timer(void)
+{
+  main_loop_assert_main_thread();
+  if (iv_timer_registered(&update_timer))
+    iv_timer_unregister(&update_timer);
+}
+
+static void
+_deinit_timer(void)
+{
+  _stop_timer();
+}
+
+void
+stats_aggregator_lock(void)
+{
+  g_static_mutex_lock(&stats_aggregator_mutex);
+  stats_aggregator_locked = TRUE;
+}
+
+void
+stats_aggregator_unlock(void)
+{
+  stats_aggregator_locked = FALSE;
+  g_static_mutex_unlock(&stats_aggregator_mutex);
+}
+
+static guint
+_stats_cluster_key_hash(const StatsClusterKey *self)
+{
+  return g_str_hash(self->id) + g_str_hash(self->instance) + self->component;
+}
+
+void
+stats_aggregator_registry_init(void)
+{
+  stats_cluster_container.clusters = g_hash_table_new_full((GHashFunc) _stats_cluster_key_hash,
+                                                           (GEqualFunc) stats_cluster_key_equal, NULL, NULL);
+  _init_timer();
+  g_static_mutex_init(&stats_aggregator_mutex);
+}
+
+void
+stats_aggregator_registry_deinit(void)
+{
+  g_hash_table_destroy(stats_cluster_container.clusters);
+  stats_cluster_container.clusters = NULL;
+  g_static_mutex_free(&stats_aggregator_mutex);
+  _deinit_timer();
+}
+
+static void
+_insert_to_table(StatsAggregator *value)
+{
+  g_hash_table_insert(stats_cluster_container.clusters, &value->key, value);
+
+  if (!iv_timer_registered(&update_timer))
+    _start_timer();
+}
+
+static gboolean
+_is_in_table(StatsClusterKey *sc_key)
+{
+  return g_hash_table_lookup(stats_cluster_container.clusters, sc_key) != NULL;
+}
+
+static void
+_remove_from_table(StatsClusterKey *sc_key)
+{
+  g_hash_table_remove(stats_cluster_container.clusters, sc_key);
+}
+
+static StatsAggregator *
+_get_from_table(StatsClusterKey *sc_key)
+{
+  return g_hash_table_lookup(stats_cluster_container.clusters, sc_key);
+}
+
+static void
+_unregister(StatsAggregator *s)
+{
+  if(s == NULL)
+    return;
+
+  stats_aggregator_untrack_counter(s);
+
+  if (stats_aggregator_is_orphaned(s) && _is_in_table(&s->key))
+    {
```

Comment:
If there is a quick client, that sends messages between two aggregation update, its value going to stay zero. As they are removed here, thus not updated. It would be nice to do here an aggregation. Of course this depends on what we would like to to with orphaned counters.

Reply:
As discussed IRL, we would like to keep the counter for longer. This change is not needed.